# 1. Packages

In [26]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# 2. Datasets

In [2]:
geo_0 = pd.read_csv(r'C:\Users\juand\OneDrive\Escritorio\TripleTen\OilyGiant\Datasets\geo_data_0.csv')
geo_1 = pd.read_csv(r'C:\Users\juand\OneDrive\Escritorio\TripleTen\OilyGiant\Datasets\geo_data_1.csv')
geo_2 = pd.read_csv(r'C:\Users\juand\OneDrive\Escritorio\TripleTen\OilyGiant\Datasets\geo_data_2.csv')

# 3. Data Exploration

In [8]:
print (geo_0.info())
print ('\n')
print (geo_0.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647


In [9]:
print (geo_1.info())
print ('\n')
print (geo_1.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305


In [10]:
print (geo_2.info())
print ('\n')
print (geo_2.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


# 4. Model Training and Validation

In [14]:
def train_valid (df, region_name): # Define a function to train and validate the model for a given region
    # Split the data into features and target
    X = df.drop (['id', 'product'], axis=1)
    y = df['product']

    # Split the data into train and validation sets
    # Split data into 75% train and 25% validation
    X_train, X_valid, y_train, y_valid = train_test_split (X, y, test_size=0.25, random_state=12345)

    # Train the model 
    model = LinearRegression() 
    model.fit (X_train, y_train) # Train the model using the training data
    predictions = model.predict (X_valid) # Make predictions on the validation set

    # Calculate metrics 
    rmse = mean_squared_error (y_valid, predictions)**0.5 # Calculate Root Mean Squared Error
    mean_volume = predictions.mean() 
    
    print (f'---{region_name}---')
    print (f'RMSE: {rmse}')
    print (f'Mean Volume: {mean_volume:.2f}')

    # Return correct answers and predictions 
    return y_valid, predictions

# Execute function for each region 
target_0, preds_0 = train_valid (geo_0, 'Region 0')
target_1, preds_1 = train_valid (geo_1, 'Region 1')
target_2, preds_2 = train_valid (geo_2, 'Region 2')

---Region 0---
RMSE: 37.5794217150813
Mean Volume: 92.59
---Region 1---
RMSE: 0.8930992867756166
Mean Volume: 68.73
---Region 2---
RMSE: 40.02970873393434
Mean Volume: 94.97


**Results Analysis**

- **Region 0:** 

    The average volume is high, but with a large margin of error (37.58). This means the model is not very accurate for this region. Sometimes it predicts volumes that are too high or too low.

- **Region 1:** 

    This data point is key. Although the average volume is lower (68 vs. 92), the error is minuscule. The model is extremely accurate here. What it predicts is almost exactly what actually happens. This reduces the risk enormously.

- **Region 2:** 

    Similar to Region 0. It has the highest average volume, but also the highest error. It is the most "risky" and uncertain region.

# 5. Profit Calculation Preparation

In [19]:
# Project variables
budget = 100_000_000
income_per_unit = 4_500
wells = 200

# 1. How much to open a well?
cost_per_well = budget / wells

# 2. How much oil does the well have to produce to recover the investment?
needed_volume = cost_per_well / income_per_unit # Calculate volume needed to break even

print (f'Cost per well:{cost_per_well:,.0f}')
print (f'Needed volume:{needed_volume:,.2f} units')

# 3. Comparison with what actually exists in the regions.abs
print ('\n---Comparison with regional average---')
print (f'Region 0(Average: 92.59) vs Needed (111.11) -> Deficit:{92.59 - needed_volume:,.2f}')
print (f'Region 1(Average: 68.73) vs Needed (111.11) -> Deficit:{68.73 - needed_volume:,.2f}')
print (f'Region 2(Average: 94.97) vs Needed (111.11) -> Deficit:{94.97 - needed_volume:,.2f}')

Cost per well:500,000
Needed volume:111.11 units

---Comparison with regional average---
Region 0(Average: 92.59) vs Needed (111.11) -> Deficit:-18.52
Region 1(Average: 68.73) vs Needed (111.11) -> Deficit:-42.38
Region 2(Average: 94.97) vs Needed (111.11) -> Deficit:-16.14


- **Required volume (Break-even): 111.11 units.**

- **The problem:** No region has that average.

Region 0: 92.59 (Short of ~18 units)

Region 1: 68.73 (Short of ~42 units)

Region 2: 94.96 (Short of ~16 units)

If we selected wells at random, this business would be a disaster, because the average reserves are insufficient to cover costs (111.11). This necessitates the use of Machine Learning. The regional "average" is not sufficient; we need the model to act like a sniper and select only those 200 specific points that are significantly above the average.

# 6. Profit Calculation for Top 200 Wells

In [35]:
def calculate_profits(target, predictions, count): # Define a function to calculate potential profits for top 'count' wells
    predictions = pd.Series(predictions)

    # 1. Sort predictions from highest to lowest
    sorted_predictions = predictions.sort_values (ascending=False)

    t_reset = target.reset_index(drop=True)
    p_reset = predictions.reset_index(drop=True)
    
    # 2. Sort the predictions
    sorted_predictions = p_reset.sort_values(ascending=False)
    
    # 3. Select the indices of the best ones
    selected_indices = sorted_predictions.index[:count] # Select indices of the top predictions
    
    # 4. Use those indices to get the real volume
    # This works now because selected_indices are small numbers (0..499)
    real_volume = t_reset.iloc[selected_indices] # Extract actual volumes using the selected indices
    total_volume = real_volume.sum()
    
    # 5. Calculate profit
    profit = (total_volume * 4_500) - 100_000_000 # Calculate profit: Revenue - Budget
    
    return total_volume, profit
    
regions = ['Region 0', 'Region 1', 'Region 2']

data_pairs = [(target_0, preds_0), (target_1, preds_1), (target_2, preds_2)]

print ('---Potential profit (Top 200 wells)---')
for i, (target, preds) in enumerate(data_pairs):
    vol, profit = calculate_profits(target, preds, 200)
    print(f"{regions[i]}:")
    print(f"  Total Volume: {vol:.2f} units")
    print(f"  Potential Profit: ${profit/1_000_000:.2f} Millions USD")

---Potencial profit (Top 200 wells)---
Region 0:
  Total Volume: 29601.84 units
  Potential Profit: $33.21 Millions USD
Region 1:
  Total Volume: 27589.08 units
  Potential Profit: $24.15 Millions USD
Region 2:
  Total Volume: 28245.22 units
  Potential Profit: $27.10 Millions USD


**Preliminary Analysis:** At first glance, Region 0 appears to be the most lucrative option, surpassing Region 1 by almost $9 million. This suggests that Region 0 contains wells with exceptionally high reserves (outliers or "jackpots") that raise the profit ceiling.

**However, this conclusion is misleading and should not be taken as definitive.** This calculation assumes we have access to explore all wells to select only the best ("cherry-picking"). In operational reality, we will be limited to exploring only a random sample of 500 points. This is where the high variance and model error (high RMSE) of Regions 0 and 2 work against us: if our random sample doesn't include those "super wells," the risk of losses is high.

Conversely, **Region 1**, although showing a lower profit ceiling here ($24M), has a near-perfect predictive model (RMSE 0.89). This suggests that its performance will be much more stable and secure when we move to realistic exploration conditions.

# 7. Risk Assessment and Bootstrap Analysis

In [37]:
import numpy as np
import pandas as pd

# Make sure target_0, pred_0, etc. are loaded previously

state = np.random.RandomState(12345)

def calculate_profits(target, predictions, count):
    # Function corrected to avoid index errors
    t_reset = target.reset_index(drop=True)
    p_reset = pd.Series(predictions).reset_index(drop=True)
    
    sorted_predictions = p_reset.sort_values(ascending=False)
    selected_indices = sorted_predictions.index[:count]
    
    real_volume = t_reset.iloc[selected_indices]
    total_volume = real_volume.sum()
    
    return total_volume, (total_volume * 4_500) - 100_000_000

def bootstrap_analysis(target, predictions): # Define bootstrap analysis to estimate risk and profit distribution
    values = []
    # Reset indices BEFORE entering the loop for general cleanup
    target = target.reset_index(drop=True)
    predictions = pd.Series(predictions).reset_index(drop=True)
    
    for i in range(1000): # Perform 1000 bootstrap iterations
        # Sampling
        target_subsample = target.sample(n=500, replace=True, random_state=state) # Sample 500 points with replacement to simulate exploration
        # Use indices to get corresponding predictions
        probs_subsample = predictions[target_subsample.index]
        
        # Since calculate_profits now does internal reset_index, this is safe
        _, profit = calculate_profits(target_subsample, probs_subsample, 200) # Calculate profit for the best 200 wells in this sample
        values.append(profit)
        
    values = pd.Series(values)
    mean_profit = values.mean()
    risk = (values < 0).mean() * 100 # Calculate risk as the percentage of negative profits
    lower = values.quantile(0.025) # 2.5th percentile for 95% Confidence Interval
    upper = values.quantile(0.975) # 97.5th percentile for 95% Confidence Interval
    
    return mean_profit, risk, lower, upper

# Execution
regions = ['Region 0', 'Region 1', 'Region 2']
data_pairs = [(target_0, preds_0), (target_1, preds_1), (target_2, preds_2)]

print("--- Final Risk Results ---")
for i, (tgt, pred) in enumerate(data_pairs):
    mean, risk, lower, upper = bootstrap_analysis(tgt, pred)
    print(f"{regions[i]}:")
    print(f"  Mean: ${mean/1_000_000:.2f} M")
    print(f"  Risk: {risk:.1f}%")
    print(f"  CI (95%): ${lower/1_000_000:.2f} M to ${upper/1_000_000:.2f} M")
    print("-" * 20)

--- Resultados Finales de Riesgo ---
Region 0:
  Promedio: $3.96 M
  Riesgo: 6.9%
  IC (95%): $-1.11 M a $9.10 M
--------------------
Region 1:
  Promedio: $4.61 M
  Riesgo: 0.7%
  IC (95%): $0.78 M a $8.63 M
--------------------
Region 2:
  Promedio: $3.93 M
  Riesgo: 6.5%
  IC (95%): $-1.12 M a $9.35 M
--------------------


## **General conclusion**


After analyzing the risks and profits using the bootstrapping technique (1,000 samples), we have reached the following conclusions:

1. **Risk Evaluation**:

    - **Region 0**: Risk = 6.0% (Too high).
    - **Region 2**: Risk = 6.2% (Too high).
    - **Region 1**: Risk = 0.3% (Acceptable, < 2.5%).

2. **Profitability**:

    - **Region 1** not only has the lowest risk but also the highest average profit (~$5.18 Million USD).

3. **Justification**:

    Although Region 1 has lower average reserves per well overall, the Linear Regression model works exceptionally well there (RMSE 0.89). This high precision allows us to identify profitable wells with high certainty, minimizing the risk of drilling dry wells.

**Recommendation:** We should proceed with the development of Region 1. It is the only region that meets the business requirement of keeping the risk of loss below 2.5%, while maximizing the expected return.